File: src/java/magma/rule/StripRule.java
package magma.rule;

import magma.error.CompileError;
import magma.input.Input;
import magma.input.Location;
import magma.input.RootInput;
import magma.node.Node;
import magma.result.Result;

/**
 * A rule implementation that strips whitespace from the input before lexing.
 * <p>
 * This rule wraps another rule and removes leading and trailing whitespace
 * from the input string before passing it to the wrapped rule for lexing.
 * During generation, it simply delegates to the wrapped rule without modification.
 * <p>
 * This rule is useful for handling input that may contain extraneous whitespace
 * that should be ignored during parsing.
 */
public record StripRule(Rule rule) implements Rule {
	@Override
	public Result<String, CompileError> generate(final Node node) {
		return this.rule.generate(node);
	}

	@Override
	public Result<Node, CompileError> lex(final Input input) {
		final String content = input.getContent(); final String strippedContent = content.strip();

		// Calculate new start and end indices after stripping
		int leadingWhitespace = content.length() - content.stripLeading().length();
		int trailingWhitespace = content.length() - content.stripTrailing().length();

		int newStartIndex = input.getStartIndex() + leadingWhitespace;
		int newEndIndex = input.getEndIndex() - trailingWhitespace;

		Location strippedLocation =
				new Location(input.getSource().getPackageSegments(), input.getSource().getName() + " (stripped)");
		final RootInput strippedInput = new RootInput(strippedContent, strippedLocation, newStartIndex, newEndIndex);
		return this.rule.lex(strippedInput);
	}
}

File: src/node/magma/rule/StripRule.ts
/*import magma.error.CompileError;*/
/*import magma.node.Node;*/
/*import magma.result.Err;*/
/*import magma.result.Ok;*/
/*import magma.result.Result;*/
/*import java.util.Optional;*/
/*public record StripRule(Rule rule) implements Rule {
	private Optional<Node> lex0(final String input) {
		return this.rule.lex(input.strip()).findValue();
	}

	private Optional<String> generate0(final Node node) {
		return this.rule.generate(node).findValue();
	}

	@Override
	public Result<Node, CompileError> lex(final String input) {
		return this.lex0(input).<Result<Node, CompileError>>map(Ok::new).orElseGet(() -> new Err<>(new CompileError()));
	}

	@Override
	public Result<String, CompileError> generate(final Node node) {
		return this.generate0(node).<Result<String, CompileError>>map(Ok::new).orElseGet(() -> new Err<>(new CompileError()));
	}
}*/
